File: CivitasJava/civitas/Tablero.java
package civitas;

import java.util.ArrayList;

public class Tablero {

    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;

    Tablero(int carcel)
    {
        this.numCasillaCarcel = (carcel >= 1) ? carcel : 1;
        this.porSalida = 0;
        this.tieneJuez = false;
        this.casillas = new ArrayList<>();
        Casilla salida=new Casilla("Salida");
        casillas.add(salida);
    }

    private boolean correcto() {
        return ((this.casillas.size() > this.numCasillaCarcel) && this.tieneJuez);
    }

    private boolean correcto(int casilla) {
        return this.correcto() && (casilla >= 0 && casilla <= this.casillas.size());
    }

    int getCarcel() {
        return this.numCasillaCarcel;
    }

    int getPorSalida() {
        int ret = this.porSalida;

        if (ret > 0) {
            this.porSalida--;
        }
        return ret;
    }

    void añadeCasilla(Casilla casilla) 
    {
        if (this.casillas.size() == this.numCasillaCarcel) {
            Casilla carcel = new Casilla("Cárcel");
            this.casillas.add(carcel);
            this.casillas.add(casilla);
        } 
        else 
        {
            this.casillas.add(casilla);

            if (this.casillas.size() == this.numCasillaCarcel) 
            {
                Casilla carcel = new Casilla("Cárcel");
                this.casillas.add(carcel);
            }
        }
    }

    void añadeJuez() {
        if (!this.tieneJuez) 
        {
            Casilla_Juez juez = new Casilla_Juez(this.numCasillaCarcel);
            this.añadeCasilla(juez);
            this.tieneJuez = true;
        }
    }

    Casilla getCasilla(int numCasilla) {
        
        Casilla ret = null;

        if (this.correcto(numCasilla)) 
        {
            ret = this.casillas.get(numCasilla);
        }

        return ret;
    }

    public ArrayList<Casilla> getCasillas()
    {
        return this.casillas;
    }

    int nuevaPosicion(int actual, int tirada) 
    {
        int nueva = -1;

        if (this.correcto()) {
            nueva = ((actual + tirada) >= this.casillas.size()) ? ((actual + tirada) % this.casillas.size())   : actual + tirada;
            if (nueva != actual + tirada) {
                this.porSalida++;
            }
        }

        return nueva;
    }

    int calcularTirada(int origen, int destino) 
    {
        return (origen - destino < 0) ? origen - destino + this.casillas.size() : origen - destino;
    }
}

File: CivitasRuby/Tablero.rb
# encoding: utf-8


module Civitas

  class Tablero
        
    attr_reader :numCasillaCarcel, :casillas

    def initialize(c_carcel)

      c_carcel>=1 ? @numCasillaCarcel=c_carcel : @numCasillaCarcel=1
      @casillas=Array.new
      @salida=Casilla.new('salida')
      @casillas << @salida      
      @porSalida=0
      @tieneJuez=false

    end

    def correcto
          
      return ((@casillas.length>@numCasillaCarcel)&&@tieneJuez)

    end

    def correcto2(numCasilla)

      return ((numCasilla<=@casillas.length) && correcto && (numCasilla>=0))

    end

    def getPorSalida
            
      ret=@porSalida

      if ret>0
        @porSalida-=1
      end
            
      return ret

    end

    def añadeCasilla(casilla)
            
      if @casillas.length==@numCasillaCarcel
        carcel=Casilla.new('Cárcel')
        @casillas << carcel
        @casillas << casilla
      else 
        @casillas << casilla
        if @casillas.length == @numCasillaCarcel
          carcel=Casilla.new('Cárcel')
          @casillas << carcel
        end    
      end  
        
    end

    def añadeJuez
        
      if !@tieneJuez
        juez=Casilla_Juez.new(@numCasillaCarcel)
        añadeCasilla(juez)
        @tieneJuez=true        
      end

    end

    def getCasilla(numCasilla)
      
      if correcto2(numCasilla)
        return @casillas[numCasilla]
      else
        return nil
      end
        
    end

    def nuevaPosicion(actual, tirada) 
        
      ret=-1

      if correcto 
                
        actual + tirada > @casillas.length-1 ? ret = ((actual + tirada) % @casillas.length) : ret = actual+tirada

        if ret != actual + tirada
          @porSalida+=1
        end

      end

      return ret

    end

    def calcularTirada(origen, destino)
        
      aux=destino- origen
      if(aux<0)
        aux=aux+@casillas.length
      end

      return aux
        
    end
      
  end

end
